src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3753,8 +3714,60 @@ void os::large_page_init() {
     return;
   }
 
-  size_t large_page_size = Linux::setup_large_page_size();
-  UseLargePages          = Linux::setup_large_page_type(large_page_size);
+  // 2) Scan OS info
+  size_t default_large_page_size = scan_default_large_page_size();
+  os::Linux::_default_large_page_size = default_large_page_size;
+  if (default_large_page_size == 0) {
+    // No large pages configured, return.
+    UseTransparentHugePages = false;
+    UseHugeTLBFS = false;
+    UseSHM = false;
+    return;
+  }
+  os::PageSizes all_large_pages = scan_multiple_page_support();
+
+  // 3) Consistency check and post-processing
+
+  // It is unclear if /sys/kernel/mm/hugepages/ and /proc/meminfo could disagree. Manually
+  // re-add the default page size to the list of page sizes to be sure.
+  all_large_pages.add(default_large_page_size);
+
+  // Populate large page sizes to _page_sizes. Add pages that
+  // are less than or equal to LargePageSizeInBytes, except when LargePageSizeInBytes=0
+  // or FLAG_IS_DEFAULT(LargePageSizeInBytes), add all sizes
```

Comment:
What we currently have in the CSR is to use the default large page size as the max if `LargePageSizeInBytes`is not set. If we first check if `LargePageSizeInBytes` have been set and update `_large_page_size` accordingly above (see comment below). I think this comment should say something like:

```suggestion
  // Populate _page_sizes with large page sizes less than or equal to
  // _large_page_size. 
```

Reply:
+1